File: games/guessing_game.hs
-- game: start with a word, status (empty) and x lives
-- turn: win || lose || guess
-- guess: take letter -> update status or lose life

takeguess status word lives =
    do  putStrLn "Enter guess: "
        guess <- getLine
        if elem (guess !! 0) word
            then taketurn (update_status guess word status) word lives
            else taketurn status word (lives - 1)

taketurn status word lives = 
    do  putStrLn (status ++ "  lives: " ++ take lives (repeat '*'))
        if (status == word) then putStrLn "You win!!"
        else if (lives == 0) then putStrLn "You lose!!"
        else takeguess status word lives

update_status :: [Char] -> String -> String -> String
update_status guess word status = 
    let jackie = zip word status
        jenny = [if x == guess !! 0 then x else y | (x,y) <- jackie]
    in jenny

game word lives = taketurn ['_' | x <- word] word lives

File: games-and-challenges/guess-a-word/guessing_game.py
def guessing_game(word,lives):
    def take_guess(status,lives):
        guess = input("take a guess: ")
        def update_status(guess):
            jackie = zip(word,status)
            jenny = map(lambda x: x[0] if x[0] == guess else x[1],jackie)
            return "".join([str(j) for j in jenny])
        if guess in word:
            return take_turn(update_status(guess),lives)
        else: 
            return take_turn(status,lives-1)    
    def take_turn(status,lives):
        print(status,lives)
        if lives == 0:
            return print("You lose")
        elif status == word:
            return print("You win!")
        else:
            return take_guess(status,lives)
    print("lets play a guessing game")
    return take_turn("{}".format("_"*len(word)),lives)

guessing_game("alphanumeric",5)

